crates/telio-starcast/src/transport.rs

```diff
@@ -0,0 +1,234 @@
+use std::{
+    net::{IpAddr, SocketAddr},
+    sync::Arc,
+};
+
+use async_trait::async_trait;
+use futures::{future::join_all, TryFutureExt};
+use ipnetwork::IpNetwork;
+use pnet_packet::{ipv4::MutableIpv4Packet, ipv6::MutableIpv6Packet};
+use telio_model::PublicKey;
+use tokio::net::UdpSocket;
+
+use telio_sockets::SocketPool;
+use telio_task::{
+    io::{wait_for_tx, Chan},
+    task_exec, Runtime, RuntimeExt, Task, WaitResponse,
+};
+use telio_utils::{telio_log_debug, telio_log_error, telio_log_warn};
+
+use crate::{
+    nat::{self, Nat, StarcastNat},
+    utils::DualIpAddr,
+};
+
+#[cfg(feature = "test-util")]
+const MULTICAST_TRANSPORT_PORT: u16 = 0;
+#[cfg(not(feature = "test-util"))]
+const MULTICAST_TRANSPORT_PORT: u16 = 13569;
+const MAX_PACKET_SIZE: usize = u16::MAX as usize;
+
+#[derive(Debug, thiserror::Error)]
+/// Starcast transport-specific errors
+pub enum Error {
+    // Could not bind to socket
+    #[error("Could not bind to socket")]
+    SocketBindError(String),
+    /// Invalid IP Packet
+    #[error("Invalid IP Packet")]
+    InvalidIpPacket,
+    /// Failed to fan out local multicast packet to some peers
+    #[error("Failed to fan out local multicast packet to peers {0:?}")]
+    FanoutFailed(Vec<PublicKey>),
+    /// Failed to nat packet
+    #[error("Failed to nat packet: {0:?}")]
+    NatError(nat::Error),
```

Comment:
Would it not make sence to also make it transparemnt, or even just add #[from] to avoid dational mappings